### DISCUSSION TITLE
Cryptic part missing elaboration in Nova 4 upgrade guide

### DISCUSSION THREAD

**Discussion by piljac1** [laravel/nova-issues#3966] [Q&A]
- Laravel Version: 9.8.1
- Nova Version: 4.1.0
- PHP Version: 8.0.16
- Database Driver & Version: MySQL 8.0.21
- Operating System and Version: macOS Monterey 12.1
- Browser type and version: Google Chrome 100.0.4896.75
- Reproduction Repository: Private

### Description:
There is a part of the upgrade guide, listed as low priority, that states the following:

> Eloquent User Provider Required
>
> Nova 4 requires the authentication user provider to be eloquent in order to resolve information regarding the currently authenticated user.
>
> In your application's default config/auth.php configuration file, the Eloquent user provider is specified and it is instructed to use the App\Models\User model when retrieving users. You may change these values within your configuration file based on the needs of your application.

Can we get some elaboration from the core team (either here or the upgrade guide) as to why is that the case and what Nova features are affected by this, so we can be aware of what we can change/disable in order to circumvent that change. We have to specifically (asked by the customer) use Auth0 as a provider in our app. It worked in Nova 3 and *seems* to work in Nova 4, but we haven't tested all features yet, including new features.

Here's how our Auth0 setup works:
```php
// auth.php

'defaults' => [
    'guard' => 'auth0',
    // ...
],

'guards' => [
    'auth0' => [
        'driver' => 'auth0',
        'provider' => 'auth0',
    ],
    // ...
],

'providers' => [
    'auth0' => [
        'driver' => 'auth0',
        'repository' => App\Auth\UserRepository::class,
    ],
    // ...
],
```

```php
<?php

namespace App\Auth;

use App\Models\User;
use Auth0\Laravel\Contract\Auth\User\Repository;
use Illuminate\Contracts\Auth\Authenticatable;

class UserRepository implements Repository
{
    /**
     * @inheritdoc
     */
    public function fromSession(array $user): ?Authenticatable
    {
        return User::makeFromAuth0UserInfo($user);
    }

    /**
     * @inheritdoc
     */
    public function fromAccessToken(array $user): ?Authenticatable
    {
        return User::makeFromAuth0UserInfo($user);
    }
}

```

P.S. `User::makeFromAuth0UserInfo($user);` creates a new App\Models\User instance, but it is not persisted in the database (no users table).

### Detailed steps to reproduce the issue on a fresh Nova installation:

- Unknown (need more info from the core team)

**Comment by crynobone** [comment]
There few reasons for this:

1. Nova pulls the current user information on each Inertia request this can easily be done with `JsonResource` on to an eloquent model.
2. When `AuthenticationSession` middleware is enabled, we are able to detect if a person updates their own password and refreshes the current request user resolver allowing the session to maintain instead of suddenly logged out.
3. Action Events still depend on an eloquent model by default.
4. Impersonating User will depend on the user to be an eloquent model.

**Reply by piljac1** [reply]
So if I understand correctly, it does not exclusively work for to the eloquent driver as stated in the upgrade guide, but it depends on whether the provider returns an eloquent model, even if using a repository to achieve that. Right?

**Reply by crynobone** [reply]
All above would require Eloquent and using anything else may fail without overriding each feature. That's why the documentation has been updated to include such restrictions.
